common/zac/handler.go

```diff
@@ -0,0 +1,120 @@
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+	https://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+package zac
+
+import (
+	"github.com/openziti/xweb/v2"
+	log "github.com/sirupsen/logrus"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	Binding = "zac"
+)
+
+type ZitiAdminConsoleFactory struct {
+}
+
+var _ xweb.ApiHandlerFactory = &ZitiAdminConsoleFactory{}
+
+func NewZitiAdminConsoleFactory() *ZitiAdminConsoleFactory {
+	return &ZitiAdminConsoleFactory{}
+}
+
+func (factory ZitiAdminConsoleFactory) Validate(*xweb.InstanceConfig) error {
+	return nil
+}
+
+func (factory ZitiAdminConsoleFactory) Binding() string {
+	return Binding
+}
+
+func (factory ZitiAdminConsoleFactory) New(_ *xweb.ServerConfig, options map[interface{}]interface{}) (xweb.ApiHandler, error) {
+	loc := options["location"]
+	if loc == nil || loc == "" {
+		log.Panic("location must be supplied in zac options")
+	}
+	indexFile := options["indexFile"]
+	if indexFile == nil || indexFile == "" {
+		indexFile = "index.html"
+	}
+	zac := &ZitiAdminConsoleHandler{
+		httpHandler: SpaHandler(loc.(string), "/"+Binding, indexFile.(string)),
+	}
+
+	return zac, nil
+}
+
+type ZitiAdminConsoleHandler struct {
+	httpHandler http.Handler
+}
+
+func (self *ZitiAdminConsoleHandler) Binding() string {
+	return Binding
+}
+
+func (self *ZitiAdminConsoleHandler) Options() map[interface{}]interface{} {
+	return nil
+}
+
+func (self *ZitiAdminConsoleHandler) RootPath() string {
+	return "/" + Binding
+}
+
+func (self *ZitiAdminConsoleHandler) IsHandler(r *http.Request) bool {
+	return strings.HasPrefix(r.URL.Path, self.RootPath()) || strings.HasPrefix(r.URL.Path, "/assets")
+}
+
+func (self *ZitiAdminConsoleHandler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	self.httpHandler.ServeHTTP(writer, request)
+}
+
+// Thanks to https://github.com/roberthodgen/spa-server
+// Serve from a public directory with specific index
+type spaHandler struct {
+	content     string // The directory from which to serve
+	contextRoot string // The context root to remove
+	indexFile   string // The fallback/default file to serve
+}
+
+// Falls back to a supplied index (indexFile) when either condition is true:
+// (1) Request (file) path is not found
+// (2) Request path is a directory
+// Otherwise serves the requested file.
+func (h *spaHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	r.URL.Path = strings.TrimPrefix(r.URL.Path, h.contextRoot)
+	p := filepath.Join(h.content, filepath.Clean(r.URL.Path))
+
+	if info, err := os.Stat(p); err != nil {
```

Comment:
This order of operations will allow an attack to use this endpoint to see what files on the host are accessible by the user hosting the controller as os.Stat is operating on `p` which isn't protected against for parent traversal or absolute paths. `p` is assumed to be a relative path.

`http.ServeFile` will respond as follows:
- path exists and is readable - http.Server will handle and then return a 400 bad request
- path exits and is not readable - redirects to index

To address this URLs should be cleaned before os.Stat is called by 
1) returning an error for any path with ".." in it
2) return an error for any path starting with `/` or `<driveLetter>:`

Reply:
It looks like you are joining to the root of the static files, so that should handle absolute root traversal. Just not `..`

Reply:
Excellent catch. I'd tried most of my tooling and went out of my way to find a utility that would actually send the `..` in the http request. curl wouldn't do it and all the browsers I tried wouldn't either. go is already using that containsDotDot function to protect people. Proof is here shown below. httppie. this HTTP response is returned from `func ServeFile(w ResponseWriter, r *Request, name string) {` in`go/pkg/mod/golang.org/toolchain@v0.0.1-go1.21.5.linux-amd64/src/net/http/fs.go`
```
$ http -v --verify=no https://192.168.1.83:1280/zac/..
GET /zac/.. HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: 192.168.1.83:1280
User-Agent: HTTPie/2.4.0



HTTP/1.1 400 Bad Request
Content-Encoding: gzip
Content-Length: 41
Content-Type: text/plain; charset=utf-8
Date: Wed, 28 Feb 2024 02:58:46 GMT
X-Content-Type-Options: nosniff

invalid URL path
```

If you still feel strongly about adding an additional check in our code, I can. I'm going to consider this resolved and I'll directly follow up wit you to see if this meets your needs or if you want the extra safety.